extensions/oplogPopulator/allocationStrategy/ImmutableConnector.js

```diff
@@ -0,0 +1,57 @@
+const AllocationStrategy = require('./AllocationStrategy');
+
+/**
+ * @class ImmutableConnector
+ *
+ * @classdesc ImmutableConnector is an allocation
+ * strategy that always requires a new connector to be
+ * created for each bucket.
+ */
+class ImmutableConnector extends AllocationStrategy {
+    /**
+     * Assign the bucket to a new connector. If a
+     * connector has retained the bucket, use it.
+     * If the bucket is already assigned to a connector,
+     * return the connector.
+     * @param {String} bucket bucket name
+     * @returns {Promise<Connector>} connector
+     */
+    async getOrAssignBucketConnector(bucket) {
+        let connector;
+        if (this._bucketsToConnectors.has(bucket)) {
+            return this._bucketsToConnectors.get(bucket);
+        } else if (this._retainedBuckets.has(bucket)) {
+            connector = this._retainedBuckets.get(bucket);
+        } else if (!connector) {
+            connector = await this._connectorsManager.addConnector();
+        }
+        this._bucketsToConnectors.set(bucket, connector);
+        await connector.addBucket(bucket);
+        return connector;
+    }
```

Comment:
Lots of duplication between the implementations, the only difference is really when we chose to reuse the connector

--> the older interface seemed better, allowing to keep only the "choice" in the Strategy.

---

Specifically for `_retainedBuckets`, it breaks layers as it is really managed in the base class (which is not really the role of an interface) but it is used here.

The cleaner way to do this is to add a separate `RetainBucket` "proxy" AllocationStrategy, which would look something like this:

```
class RetainBuckets extends AllocationStrategy {
    constructor(strategy, params) {
        this._strategy = strategy;
        this._retainedBuckets = new Map();

        params.connectorsManager.on('update', () => {
            this._retainedBuckets.forEach(bucket => {
                if (!this._bucketsToConnectors.has(bucket)) {
                    this._retainedBuckets.delete(bucket);
                }
            });
            this._retainedBuckets.clear();
        });
    }

    getBucketConnector(bucket) {
        if (this._retainedBuckets.has(bucket)) {
            return this._retainedBuckets.get(bucket);
        }

        return this._strategy.getBucketConnector(bucket)
    }

    // the interface expects to return TRUE if the connector must be removed, or FALSE to keep it
    removeBucketFromConnector(bucket, connector) {   
        const res = this._strategy.removeBucketFromConnector(bucket, connector);
        if res {
            this._retainedBuckets.set(bucket, connector);
        }
        return res;
    }
}
```

This way we mananged the retained buckets in a single place, and the "retaining" code is neither duplicated nor intermingled with the rest of the code.